src/helpers/proposed_network.rs

```diff
@@ -0,0 +1,198 @@
+#![allow(dead_code, clippy::mutable_key_type)]
+
+use crate::{
+    helpers::{network::MessageChunks, ByteArrStream, Role},
+    protocol::QueryId,
+};
+use futures::Stream;
+use hyper::Uri;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use tokio::sync::oneshot;
+
+#[derive(Debug, thiserror::Error)]
+pub enum NetworkEventError {
+    #[error("event {event_name} failed to respond to callback for query_id {}", .query_id.as_ref())]
+    CallbackFailed {
+        event_name: &'static str,
+        query_id: QueryId,
+    },
+}
+
+pub struct SendMessageData {
+    pub chunks: MessageChunks,
+}
+
+impl SendMessageData {
+    pub fn new(chunks: MessageChunks) -> Self {
+        Self { chunks }
+    }
+}
+
+/// Events sent within the context of executing a query
+pub enum RingEvent {
+    SendMessage(SendMessageData),
+}
+
+pub trait Ring: Sync {
+    /// Type of the channel that is used to send messages to other helpers
+    type Sink: futures::Sink<RingEvent, Error = NetworkEventError> + Send + Unpin + 'static;
+    type MessageStream: Stream<Item = RingEvent> + Send + Unpin + 'static;
+
+    /// Returns a sink that accepts data to be sent to other helper parties.
+    fn sink(&self) -> Self::Sink;
+
+    /// Returns a stream to receive messages that have arrived from other helpers. Note that
+    /// some implementations may panic if this method is called more than once.
+    fn recv_stream(&self) -> Self::MessageStream;
+}
+
+pub struct CreateQueryData {
+    callback: oneshot::Sender<QueryId>,
+}
+
+impl CreateQueryData {
+    #[must_use]
+    pub fn new(callback: oneshot::Sender<QueryId>) -> Self {
+        CreateQueryData { callback }
+    }
+    pub fn respond(self, query_id: QueryId) -> Result<(), NetworkEventError> {
+        self.callback
+            .send(query_id)
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "CreateQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct PrepareQueryData {
+    pub query_id: QueryId,
+    callback: oneshot::Sender<()>,
+}
+
+impl PrepareQueryData {
+    #[must_use]
+    pub fn new(query_id: QueryId, callback: oneshot::Sender<()>) -> Self {
+        PrepareQueryData { query_id, callback }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "PrepareQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct StartMulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
```

Comment:
that should be part of prepare query API as well as field_type. I think we should also make query type enum available, so followers would know the data format

Reply:
what will prepare query do with this info other than just store it until this step?

i'm not sure i follow what you mean about making the query type enum available. Wouldn't followers receive the `StartMul` command, which will mean they know the data format based on that?

Reply:
we need to keep track of running queries on followers too, so letting followers know about query parameters as part of prepare query API looks logical to me. data upload just carries query id and data associated with it.


wrt to query type, I wasn't sure whether we need this information or no. Clients certainly have to tell us which query type they want to run, do we need to keep the query type in query metadata? does not seem that we strictly need to, but it feels a bit weird to trust leader to correctly handle this without checking it. For example, lets say there is a bit flip in query type that caused leader to believe it is running malicious IPA and followers running sem-honest. The input format would be the same, but result will be a complete garbage. It is probably out of my paranoia again, if cost of having it in metadata is high, we don't have to implement it

Reply:
gotcha. so we want to optimize for upload being as simple as possible. i will adjust for that.

as for query type, sure i can add that. I didn't see any enum to choose between semi_honest and malicious, so i created one. but will remove if there's something that already exists